### ISSUE TITLE
qRegisterMetatype replacement

### ISSUE THREAD

**Issue by arknable** [therecipe/qt#115]
Hi,
I'm trying to find out what is the replacement of `qRegisterMetatype` in core but can't find it. What does the equivalent function for this ?

Thanks.

**Comment by therecipe** [comment]
Hey

Currently only class member functions are supported.
I will look into this.

**Comment by arknable** [comment]
Hi. ok then, thanks for the information. I'll keep this open just for reminder if you don't mind :)

**Comment by therecipe** [comment]
I'm already working on it ;)
I planned to make both `qRegisterMetatype` functions available to all `qtmoc` generated subclasses.

Is that sufficient?
How do you plan to use the function?
Just to dynamically create instances in QML?

**Comment by arknable** [comment]
ah wonderful, big thanks. I need it to return custom class to QML. that macro wrapper will suffice, I just need to register the type so QML can recognize it.

**Comment by therecipe** [comment]
The `qRegisterMetaType` functions should be available now with https://github.com/therecipe/qt/commit/02751dcb9cd9c731fcc4c25ad08418c89208508b for all "moc" classes.

Also the [metatype macro](http://doc.qt.io/qt-5/qmetatype.html#Q_DECLARE_METATYPE) is always used.

Hope that works :)

**Comment by arknable** [comment]
I tried to test this several times, but my custom object still doesn't recognized. is there any particular step that I need to do for this to work using the binding ?. The custom class is basically as simple as this:
```
type NativeResult struct {
          core. QObject
}
```

and I use that struct as return value of a slot. but still got error `Error: Unknown method return type: NativeResult*`

**Comment by arknable** [comment]
ah sorry I don't know why I closed this.

**Comment by arknable** [comment]
Hi,

I tried this again using your sample for QML engine. it's working for context property, thanks. But I still can't get my custom class as function return type to QML.

```
type NativeResult struct {
	core.QObject

	_ func() *NativeResult `slot:instance`
}

func (l *Logic) NewResult() *NativeResult {
	result := NewNativeResult(nil)
	result.QRegisterMetaType()
	result.ConnectInstance(func() *NativeResult {
		return result
	})
	return result
}
```

and on other place I'm using the type as return value in a slot function to be used in QML.

```
func (l *Logic) foo() *NativeResult {
	result := l.NewResult()
	// do something
	return result
}
```

the slot is defined in the bridge:

```
type QmlBridge struct {
	core.QObject

	_ func(address string) *NativeResult       `slot:"unblock"`
	_ func() *QmlBridge                        `slot:"instance"`
}
```

am I missing something ?

thanks.

**Comment by therecipe** [comment]
Hey

I could reproduce it.
Somehow you can't return a `*NativeResult` from the `QmlBridge`.
You can only return a `*NativeResult` from the `NativeResult`,  a `*QmlBride` from the `QmlBridge`, ...
I will try to fix it.


However it works and is recognized as a NativeResult in QML if I replace
```
_ func(address string) *NativeResult       `slot:"unblock"`
```

with

```
_ func(address string) *core.QObject       `slot:"unblock"`
```

and get the `*core.QObject` pointer with something like `foo().QObject_PTR()` in `unblock`.

Hope that helps.

**Comment by therecipe** [comment]
Found it in the docs:
> To use the type T in queued signal and slot connections, qRegisterMetaType<T>() must be called before the first connection is established.

With the design you used above `(*QmlBridge).ConnectUnblock()` is called before `result.QRegisterMetaType()` and therefore `*NativeResult` not recognized by the `QmlBridge`.

I made the `QRegisterMetaType` functions static so that it can now be used like this:

`QmlBridge_QRegisterMetaType()` and `NativeResult_QRegisterMetaType()` before the call to 
`(*QmlBridge).ConnectUnblock(func(address string) *NativeResult`

I pushed the changes with https://github.com/therecipe/qt/commit/c9176cdd9a8fa666f6e49e47ed91d287c70a5300

(You probably only need to re-compile qtmoc to get the new functions)

**Comment by arknable** [comment]
Hi,

I tried this and the static function seems not recognized:

```
func (l *Logic) NewBridge() *QmlBridge {
	output := NewQmlBridge(nil)
	output.QRegisterMetaType()
	NativeResult_QRegisterMetaType()
	output.ConnectBlacklist_Unblock(l.Unblock)
	output.ConnectInstance(func() *QmlBridge {
		return output
	})

	return output
}
```

error was: `undefined: NativeResult_QRegisterMetaType`

I also rebuild the binding. am I did that correctly ?

**Comment by therecipe** [comment]
Yes, re-building the binding should have worked as well.
Did you run `qtmoc` in the project folder?
After that the moc* files should be re-generated and the `NativeResult_QRegisterMetaType()` function should be available.

**Comment by arknable** [comment]
yes, I did. also tried to remove all moc files run `qtmoc`. let me try again with rebuilt binding

**Comment by arknable** [comment]
Hey, 
So I tried to do this in a simpler code and `NativeResult_QMetaType()` doesn't recognized. But I re-read again your comment above and I think I missed few lines and tried your suggestion to use `core.QObject` instead and it works. so at least the logic function can return strongly-typed result and passed as generic QObject pointer to QML using `QObject_PTR()`. My test code is as follows.

`main.go`
```
package main

import (
	"os"

	"github.com/therecipe/qt/core"
	"github.com/therecipe/qt/gui"
	"github.com/therecipe/qt/qml"
)

type NativeResult struct {
	core.QObject
	message string
	error   bool

	_ func() string `slot:"Message"`
	_ func() bool   `slot:"HasError"`
}

func NewResult(error bool, message string) *NativeResult {
	var result = NewNativeResult(nil)
	result.message = message
	result.error = error
	result.ConnectMessage(func() string { return result.message })
	result.ConnectHasError(func() bool { return result.error })
	return result
}

type Controller struct {
	core.QObject
	_ func() *core.QObject `slot:"GreetingMessage"`
}

func (c *Controller) getGreetingMessage() *core.QObject {
	var result = NewResult(false, "Greeting from Go 123 !!")
	return result.QObject_PTR()
}

func main() {
	var controller = NewController(nil)
	controller.QRegisterMetaType()
	controller.ConnectGreetingMessage(controller.getGreetingMessage)

	gui.NewQGuiApplication(len(os.Args), os.Args)
	engine := qml.NewQQmlApplicationEngine(nil)
	engine.RootContext().SetContextProperty("_gController", controller)
	engine.Load(core.NewQUrl3("qrc:///qml/main.qml", 0))
	gui.QGuiApplication_Exec()
}
```

`main.qml`
```
import QtQuick 2.7
import QtQuick.Window 2.2
import QtQuick.Controls 2.0
import QtQuick.Layouts 1.1

ApplicationWindow {
    id: window
    visible: true
    width: 800
    minimumWidth: 800
    height: 500
    minimumHeight: 500

    Component.onCompleted: {
        console.log("Message: " + _gController.GreetingMessage().Message());
        console.log("Error?: " + _gController.GreetingMessage().HasError());
    }
}
```

the output is correct:
```
qml: Message: Greeting from Go 123 !!
qml: Error?: false
```

**Comment by arknable** [comment]
so I guess we can close this issue.

**Comment by therecipe** [comment]
Yes, using the generic core.QObject class should work as well.
But it's still strange that the `NativeResult_QRegisterMetaType()` function is not recognized.

I had a similar problem some time ago, where I needed to remove the old `*.a` files in `$GOPATH/pkg/github.com/therecipe/qt/` first before re-building the binding.
Maybe that could help.

**Comment by arknable** [comment]
hmm, probably that's the problem. I'll try, thanks.
